keras_hub/src/models/retinanet/non_max_supression.py

```diff
@@ -0,0 +1,586 @@
+                ops.cast(
+                    ops.linspace(
+                        idx * tile_size, (idx + 1) * tile_size - 1, tile_size
+                    ),
+                    "int32",
+                ),
+                axis=0,
+            ),
+            axis=-1,
+        )
+        box_slice = ops.take_along_axis(boxes, slice_index, axis=1)
+        _, box_slice, _, _ = ops.while_loop(
+            lambda _boxes, _box_slice, _threshold, inner_idx: inner_idx < idx,
+            cross_suppression_func,
+            [boxes, box_slice, iou_threshold, ops.array(0)],
+        )
+
+        # Iterates over the current tile to compute self-suppression.
+        iou = _bbox_overlap(box_slice, box_slice)
+        mask = ops.expand_dims(
+            ops.reshape(ops.arange(tile_size), [1, -1])
+            > ops.reshape(ops.arange(tile_size), [-1, 1]),
+            0,
+        )
+        iou *= ops.cast(ops.logical_and(mask, iou >= iou_threshold), iou.dtype)
+        suppressed_iou, _, _, _ = ops.while_loop(
+            lambda _iou, loop_condition, _iou_sum, _: loop_condition,
+            _self_suppression,
+            [iou, ops.array(True), ops.sum(iou, [1, 2]), iou_threshold],
+        )
+        suppressed_box = ops.sum(suppressed_iou, 1) > 0
+        box_slice *= ops.expand_dims(
+            1.0 - ops.cast(suppressed_box, box_slice.dtype), 2
+        )
+
+        # Uses box_slice to update the input boxes.
+        mask = ops.reshape(
+            ops.cast(ops.equal(ops.arange(num_tiles), idx), boxes.dtype),
+            [1, -1, 1, 1],
+        )
+        boxes = ops.tile(
+            ops.expand_dims(box_slice, 1), [1, num_tiles, 1, 1]
+        ) * mask + ops.reshape(boxes, [batch_size, num_tiles, tile_size, 4]) * (
+            1 - mask
+        )
+        boxes = ops.reshape(boxes, [batch_size, -1, 4])
+
+        # Updates output_size.
+        output_size += ops.cast(
+            ops.sum(ops.any(box_slice > 0, [2]), [1]), "int32"
+        )
+    return boxes, iou_threshold, output_size, idx + 1
+
+
+def mask_invalid_detections(bounding_boxes):
+    """masks out invalid detections with -1s.
+
+    This utility is mainly used on the output of non-max suppression operations.
+    The output of non-max-suppression contains all the detections, even invalid
+    ones. Users are expected to use `num_detections` to determine how many boxes
+    are in each image.
+
+    In contrast, KerasCV expects all bounding boxes to be padded with -1s.
+    This function uses the value of `num_detections` to mask out
+    invalid boxes with -1s.
+
+    Args:
+        bounding_boxes: a dictionary complying with KerasCV bounding box format.
+            In addition to the normal required keys, these boxes are also
+            expected to have a `num_detections` key.
+        output_ragged: whether to output RaggedTensor based bounding
+            boxes.
+    Returns:
+        bounding boxes with proper masking of the boxes according to
```

Comment:
this seems out of date right? we are just assuming padding boxes?

Reply:
Yes, there was a functionality ragify but removed that logic as we intend to throw if there are no ragged tensors as well for input.

Forgot to remove the arg description, changed it now.